andi/andi.py

```diff
@@ -1,62 +1,318 @@
     For each argument of the ``func`` return a list of possible types.
+    Non annotated arguments are also returned with an empty list of possible
+    types.
     """
     globalns = get_globalns(func)
     annotations = get_type_hints(func, globalns)
+    for name in get_unannotated_params(func, annotations):
+        annotations[name] = None
     annotations.pop('return', None)
-    annotations.pop('cls', None)  # FIXME: pop first argument of methods
+    annotations.pop('self', None)  # FIXME: pop first argument of methods
+    annotations.pop('cls', None)
     res = {}
     for key, tp in annotations.items():
         if is_union(tp):
             res[key] = get_union_args(tp)
         else:
-            res[key] = [tp]
+            res[key] = [] if tp is None else [tp]
     return res
 
 
-def to_provide(
-        arguments_or_func: Union[
-            Callable,
-            Dict[str, List[Optional[Type]]]
-        ],
-        can_provide: Union[
-            Container[Type],
-            Callable[[Optional[Type]], bool]
-        ]
-        ) -> Dict[str, Optional[Type]]:
+ContainerOrCallableType = Union[
+    Container[Callable],
+    Callable[[Callable], bool]
+]
+
+
+PlanMapping = MutableMapping[Type, Dict[str, Type]]
+
+
+class Plan(PlanMapping):
+    """
+    The plan resultant of executing the ``plan`` function.
     """
-    Return a dictionary ``{argument_name: type}`` with types which should
-    be provided.
 
-    ``arguments_or_func`` should be either a callable, or
-    a result of andi.inspect call for a callable.
+    def __init__(self):
+        self._plan = OrderedDict()
+
+    def __getitem__(self, item):
+        return self._plan.__getitem__(item)
+
+    def __setitem__(self, key, value):
+        return self._plan.__setitem__(key, value)
+
+    def __delitem__(self, key):
+        return self._plan.__delitem__(key)
+
+    def __iter__(self):
+        return self._plan.__iter__()
+
+    def __len__(self):
+        return self._plan.__len__()
+
+    def __str__(self):
+        return self._plan.__str__()
+
+    def __repr__(self):
+        return self._plan.__repr__()
+
+    @property
+    def dependencies(self) -> PlanMapping:
+        """
+        The plan for build the dependencies of the last task of the plan.
+        Useful when it is known that the last task of the plan could be
+        incomplete, that is, not all dependencies for the last task could be
+        resolved. In such a case is convenient to execute the plan
+        for the dependencies and the have a custom execution for the
+        last task of the plan.
+        """
+        return OrderedDict(list(self.items())[:-1])
+
+    @property
+    def final_arguments(self) -> Dict[str, Type]:
+        """
+         The argument names and its types for those arguments for
+         which it was possible to fulfil the dependencies.
+        """
+        _, params = list(self.items())[-1]
+        return params
+
+
+def plan(class_or_func: Callable, *,
```

Comment:
Using `Callable` as type because any class is also a Callable.

Reply:
a good catch